File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/ClusterMarkerHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.Marker
import com.naver.maps.map.overlay.Overlay
import io.flutter.plugin.common.MethodChannel

internal interface ClusterMarkerHandler {
    fun handleClusterMarker(
        marker: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (marker as Marker).let { m ->
        when (method) {
            "lSyncClusterMarker" -> syncClusterMarker(m, arg!!, result::success)
            else -> result.notImplemented()
        }
    }

    fun syncClusterMarker(marker: Marker, rawClusterMarker: Any, success: (Any?) -> Unit)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/ClusterMarkerHandler.swift
import NMapsMap
import Flutter

internal protocol ClusterMarkerHandler {
    func syncClusterMarker(_ marker: NMFMarker, rawClusterMarker: Any, success: (Any?) -> ())
}

internal extension ClusterMarkerHandler {
    func handleClusterMarker(
        marker: NMFOverlay, 
        method: String,
        arg: Any?,
        result: @escaping FlutterResult) {
        let marker = marker as! NMFMarker
        
        switch method {
        case syncClusterMarkerName:
            syncClusterMarker(marker, rawClusterMarker: arg!, success: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}

internal let syncClusterMarkerName = "lSyncClusterMarker"
